File: backend/scripts/check_translations.py
#!/usr/bin/env python3

"""
Backend Translation Completeness Checker

Checks that all translation keys exist in both Dutch (nl) and English (en) languages.
Reports missing keys, extra keys, and provides statistics.
"""

import json
import os
import sys
from pathlib import Path
from typing import Dict, List, Set, Tuple

# Colors for console output
class Colors:
    RESET = '\033[0m'
    RED = '\033[31m'
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    CYAN = '\033[36m'

# Translation file paths
BACKEND_DIR = Path(__file__).parent.parent
TRANSLATIONS_DIR = BACKEND_DIR / 'translations'
LANGUAGES = ['nl', 'en']

def load_json(file_path: Path) -> Dict:
    """Load JSON file"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception as e:
        print(f"{Colors.RED}Error loading {file_path}: {e}{Colors.RESET}")
        return None

def get_all_keys(obj: Dict, prefix: str = '') -> List[str]:
    """Get all keys from nested dictionary"""
    keys = []
    
    for key, value in obj.items():
        full_key = f"{prefix}.{key}" if prefix else key
        
        if isinstance(value, dict):
            keys.extend(get_all_keys(value, full_key))
        else:
            keys.append(full_key)
    
    return keys

def compare_keys(keys1: List[str], keys2: List[str]) -> Tuple[List[str], List[str]]:
    """Compare two sets of keys"""
    set1 = set(keys1)
    set2 = set(keys2)
    
    missing = [k for k in keys1 if k not in set2]
    extra = [k for k in keys2 if k not in set1]
    
    return missing, extra

def check_translations() -> bool:
    """Check all backend translations"""
    print(f"{Colors.BLUE}")
    print("╔═══════════════════════════════════════════════════════════╗")
    print("║      Backend Translation Completeness Checker            ║")
    print("║         Checking Dutch (nl) vs English (en)               ║")
    print("╚═══════════════════════════════════════════════════════════╝")
    print(f"{Colors.RESET}\n")
    
    # Check if translations directory exists
    if not TRANSLATIONS_DIR.exists():
        print(f"{Colors.YELLOW}No translations directory found at {TRANSLATIONS_DIR}{Colors.RESET}")
        print(f"{Colors.YELLOW}Backend translations are optional (using Flask-Babel){Colors.RESET}")
        return True
    
    # Find all translation files
    nl_files = list((TRANSLATIONS_DIR / 'nl').glob('*.json')) if (TRANSLATIONS_DIR / 'nl').exists() else []
    en_files = list((TRANSLATIONS_DIR / 'en').glob('*.json')) if (TRANSLATIONS_DIR / 'en').exists() else []
    
    if not nl_files and not en_files:
        print(f"{Colors.YELLOW}No translation files found{Colors.RESET}")
        print(f"{Colors.YELLOW}Backend uses Flask-Babel for translations{Colors.RESET}")
        return True
    
    # Get all unique file names
    all_files = set([f.name for f in nl_files] + [f.name for f in en_files])
    
    has_errors = False
    all_stats = []
    
    for filename in sorted(all_files):
        print(f"\n{Colors.CYAN}Checking file: {filename}{Colors.RESET}")
        print("=" * 60)
        
        nl_path = TRANSLATIONS_DIR / 'nl' / filename
        en_path = TRANSLATIONS_DIR / 'en' / filename
        
        # Check if both files exist
        if not nl_path.exists():
            print(f"{Colors.RED}✗ Missing Dutch file: {nl_path}{Colors.RESET}")
            has_errors = True
            continue
        
        if not en_path.exists():
            print(f"{Colors.RED}✗ Missing English file: {en_path}{Colors.RESET}")
            has_errors = True
            continue
        
        # Load translations
        nl_data = load_json(nl_path)
        en_data = load_json(en_path)
        
        if nl_data is None or en_data is None:
            has_errors = True
            continue
        
        # Get all keys
        nl_keys = get_all_keys(nl_data)
        en_keys = get_all_keys(en_data)
        
        # Compare keys
        missing, extra = compare_keys(nl_keys, en_keys)
        
        # Report results
        if not missing and not extra:
            print(f"{Colors.GREEN}✓ All keys match! ({len(nl_keys)} keys){Colors.RESET}")
        else:
            has_errors = True
            
            if missing:
                print(f"\n{Colors.RED}Missing in English ({len(missing)}):{ Colors.RESET}")
                for key in missing:
                    print(f"  - {key}")
            
            if extra:
                print(f"\n{Colors.YELLOW}Extra in English ({len(extra)}):{ Colors.RESET}")
                for key in extra:
                    print(f"  - {key}")
        
        all_stats.append({
            'file': filename,
            'nl': len(nl_keys),
            'en': len(en_keys),
            'missing': len(missing),
            'extra': len(extra)
        })
    
    # Print summary
    if all_stats:
        print(f"\n{Colors.CYAN}Summary{Colors.RESET}")
        print("=" * 60)
        print("\nFile Statistics:")
        print("┌─────────────────────┬────────┬────────┬─────────┬───────┐")
        print("│ File                │ NL     │ EN     │ Missing │ Extra │")
        print("├─────────────────────┼────────┼────────┼─────────┼───────┤")
        
        total_nl = 0
        total_en = 0
        total_missing = 0
        total_extra = 0
        
        for stat in all_stats:
            print(f"│ {stat['file'][:19].ljust(19)} │ {str(stat['nl']).rjust(6)} │ {str(stat['en']).rjust(6)} │ {str(stat['missing']).rjust(7)} │ {str(stat['extra']).rjust(5)} │")
            
            total_nl += stat['nl']
            total_en += stat['en']
            total_missing += stat['missing']
            total_extra += stat['extra']
        
        print("├─────────────────────┼────────┼────────┼─────────┼───────┤")
        print(f"│ {Colors.CYAN}TOTAL{Colors.RESET}               │ {str(total_nl).rjust(6)} │ {str(total_en).rjust(6)} │ {str(total_missing).rjust(7)} │ {str(total_extra).rjust(5)} │")
        print("└─────────────────────┴────────┴────────┴─────────┴───────┘")
        
        print(f"\nTotal translation keys: {total_nl} (Dutch), {total_en} (English)")
        
        if total_missing > 0:
            print(f"{Colors.RED}Missing translations: {total_missing}{Colors.RESET}")
        
        if total_extra > 0:
            print(f"{Colors.YELLOW}Extra translations: {total_extra}{Colors.RESET}")
    
    # Return result
    if has_errors:
        print(f"\n{Colors.RED}✗ Translation check failed!{Colors.RESET}")
        return False
    else:
        print(f"\n{Colors.GREEN}✓ All translations are complete!{Colors.RESET}")
        return True

if __name__ == '__main__':
    success = check_translations()
    sys.exit(0 if success else 1)

File: frontend/scripts/check-translations.js
#!/usr/bin/env node

/**
 * Translation Completeness Checker
 * 
 * Checks that all translation keys exist in both Dutch (nl) and English (en) languages.
 * Reports missing keys, extra keys, and provides statistics.
 */

const fs = require('fs');
const path = require('path');

// Colors for console output
const colors = {
  reset: '\x1b[0m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  cyan: '\x1b[36m',
};

// Translation file paths
const LOCALES_DIR = path.join(__dirname, '..', 'src', 'locales');
const LANGUAGES = ['nl', 'en'];
const NAMESPACES = ['common', 'auth', 'reports', 'str', 'banking', 'admin', 'errors', 'validation'];

/**
 * Load JSON file
 */
function loadJSON(filePath) {
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    return JSON.parse(content);
  } catch (error) {
    console.error(`${colors.red}Error loading ${filePath}: ${error.message}${colors.reset}`);
    return null;
  }
}

/**
 * Get all keys from nested object
 */
function getAllKeys(obj, prefix = '') {
  const keys = [];
  
  for (const key in obj) {
    const fullKey = prefix ? `${prefix}.${key}` : key;
    
    if (typeof obj[key] === 'object' && obj[key] !== null && !Array.isArray(obj[key])) {
      keys.push(...getAllKeys(obj[key], fullKey));
    } else {
      keys.push(fullKey);
    }
  }
  
  return keys;
}

/**
 * Compare two sets of keys
 */
function compareKeys(keys1, keys2) {
  const set1 = new Set(keys1);
  const set2 = new Set(keys2);
  
  const missing = keys1.filter(key => !set2.has(key));
  const extra = keys2.filter(key => !set1.has(key));
  
  return { missing, extra };
}

/**
 * Check translations for a namespace
 */
function checkNamespace(namespace) {
  console.log(`\n${colors.cyan}Checking namespace: ${namespace}${colors.reset}`);
  console.log('='.repeat(60));
  
  const translations = {};
  let hasErrors = false;
  
  // Load translations for all languages
  for (const lang of LANGUAGES) {
    const filePath = path.join(LOCALES_DIR, lang, `${namespace}.json`);
    
    if (!fs.existsSync(filePath)) {
      console.log(`${colors.red}✗ Missing file: ${filePath}${colors.reset}`);
      hasErrors = true;
      continue;
    }
    
    translations[lang] = loadJSON(filePath);
    
    if (!translations[lang]) {
      hasErrors = true;
      continue;
    }
  }
  
  // If any language is missing, skip comparison
  if (Object.keys(translations).length !== LANGUAGES.length) {
    return { hasErrors: true, stats: null };
  }
  
  // Get all keys for each language
  const keys = {};
  for (const lang of LANGUAGES) {
    keys[lang] = getAllKeys(translations[lang]);
  }
  
  // Compare Dutch (reference) with English
  const { missing, extra } = compareKeys(keys.nl, keys.en);
  
  // Report results
  if (missing.length === 0 && extra.length === 0) {
    console.log(`${colors.green}✓ All keys match! (${keys.nl.length} keys)${colors.reset}`);
  } else {
    hasErrors = true;
    
    if (missing.length > 0) {
      console.log(`\n${colors.red}Missing in English (${missing.length}):${colors.reset}`);
      missing.forEach(key => console.log(`  - ${key}`));
    }
    
    if (extra.length > 0) {
      console.log(`\n${colors.yellow}Extra in English (${extra.length}):${colors.reset}`);
      extra.forEach(key => console.log(`  - ${key}`));
    }
  }
  
  return {
    hasErrors,
    stats: {
      namespace,
      nl: keys.nl.length,
      en: keys.en.length,
      missing: missing.length,
      extra: extra.length
    }
  };
}

/**
 * Main function
 */
function main() {
  console.log(`${colors.blue}
╔═══════════════════════════════════════════════════════════╗
║         Translation Completeness Checker                  ║
║         Checking Dutch (nl) vs English (en)               ║
╚═══════════════════════════════════════════════════════════╝
${colors.reset}`);
  
  const allStats = [];
  let totalErrors = false;
  
  // Check each namespace
  for (const namespace of NAMESPACES) {
    const result = checkNamespace(namespace);
    
    if (result.hasErrors) {
      totalErrors = true;
    }
    
    if (result.stats) {
      allStats.push(result.stats);
    }
  }
  
  // Print summary
  console.log(`\n${colors.cyan}Summary${colors.reset}`);
  console.log('='.repeat(60));
  
  if (allStats.length > 0) {
    console.log('\nNamespace Statistics:');
    console.log('┌─────────────┬────────┬────────┬─────────┬───────┐');
    console.log('│ Namespace   │ NL     │ EN     │ Missing │ Extra │');
    console.log('├─────────────┼────────┼────────┼─────────┼───────┤');
    
    let totalNL = 0;
    let totalEN = 0;
    let totalMissing = 0;
    let totalExtra = 0;
    
    allStats.forEach(stat => {
      console.log(
        `│ ${stat.namespace.padEnd(11)} │ ${String(stat.nl).padStart(6)} │ ${String(stat.en).padStart(6)} │ ${String(stat.missing).padStart(7)} │ ${String(stat.extra).padStart(5)} │`
      );
      
      totalNL += stat.nl;
      totalEN += stat.en;
      totalMissing += stat.missing;
      totalExtra += stat.extra;
    });
    
    console.log('├─────────────┼────────┼────────┼─────────┼───────┤');
    console.log(
      `│ ${colors.cyan}TOTAL${colors.reset}       │ ${String(totalNL).padStart(6)} │ ${String(totalEN).padStart(6)} │ ${String(totalMissing).padStart(7)} │ ${String(totalExtra).padStart(5)} │`
    );
    console.log('└─────────────┴────────┴────────┴─────────┴───────┘');
    
    console.log(`\nTotal translation keys: ${totalNL} (Dutch), ${totalEN} (English)`);
    
    if (totalMissing > 0) {
      console.log(`${colors.red}Missing translations: ${totalMissing}${colors.reset}`);
    }
    
    if (totalExtra > 0) {
      console.log(`${colors.yellow}Extra translations: ${totalExtra}${colors.reset}`);
    }
  }
  
  // Exit with appropriate code
  if (totalErrors) {
    console.log(`\n${colors.red}✗ Translation check failed!${colors.reset}`);
    process.exit(1);
  } else {
    console.log(`\n${colors.green}✓ All translations are complete!${colors.reset}`);
    process.exit(0);
  }
}

// Run the script
main();
